### DISCUSSION TITLE
What is the reason for the error? ERROR: Solver (ipopt) returned non-zero return code (1)

### DISCUSSION THREAD

**Discussion by JieZhu6** [coin-or/Ipopt#701] [Q&A] [answered]
I'm getting an error while using ipopt to solve a model that contains constraints on the tanh function. The error log is as follows:
```
ERROR: Solver (ipopt) returned non-zero return code (1)
ERROR: Solver log: Ipopt 3.14.12:

******************************************************************************
 This program contains Ipopt, a library for large-scale nonlinear
optimization.
 Ipopt is released as open source code under the Eclipse Public License
 (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

This is Ipopt version 3.14.12, running with linear solver MUMPS 5.5.1.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:     3249
Number of nonzeros in Lagrangian Hessian.............:      745

Error evaluating "var =" definition -1: can't evaluate tanh'(1152.63).
```
I don't know what the problem is that's causing this

**Comment by svigerske** [comment, marked as answer]
The "Error evaluating ..." line seems to come from the AMPL solver library (ASL) code when handling a defined variable. The derivative of tanh is essentially 0 at this point, but I guess while computing it, it tries to evaluate exp(1152), which runs into a floating-point overflow. The AMPL developers should know better what is happening there.

You may want to revise your starting point or model to avoid that tanh' needs to be evaluated at such a large value.

**Reply by JieZhu6** [reply]
Here are the constraints I've troubleshot that are causing the error to be reported. Where gmm() is a function containing tanh and m.Bus_V is a variable ranging from 0.95 to 1.05. What puzzles me is that when I remove the constants 1.05 and 0.95, the model is solvable, and as soon as I add these two constants it reports an error. （I have ignored the other constraints in the model in order to troubleshoot the error.）
def joint_V(m): 
   i,j = 11,15
   v_upcon = gmm(1.05-m.Bus_V[i,j],v_mean[i,j],v_std[i,j],weight)
   v_downcon = gmm(0.95-m.Bus_V[i,j+1],v_mean[i,j],v_std[i,j],weight)
   return v_upcon - v_downcon >= 0.95
    The following constraint is the one that doesn't report an error, and I didn't add any other constraints in the model in order to check the problem.
def joint_V(m): 
   i,j = 11,15
   v_upcon = gmm(m.Bus_V[i,j],v_mean[i,j],v_std[i,j],weight)
   v_downcon = gmm(m.Bus_V[i,j+1],v_mean[i,j],v_std[i,j],weight)
   return v_upcon - v_downcon >= 0.95
